### PR TITLE
DAOS-10681 placement: fix the shard colocate issue

### PR DESCRIPTION
**Pull request by wangdi1**
DAOS-10681 placement: fix the shard colocate issue
    
Fix in get_target() to make sure shards within the
same group will not in the same failure domain, though
it might put multiple shards from different group in
the same VOS target.
    
Add object layout upgrade path for this layout fix.
    
Add dst_reclaim_ver to avoid deleting wrong OID once
rebuild failes.
    
Remove unnecessary assert once we allow put multiple objects in
the same target, i.e. any objects can be rebuilt.
    
Add jump_map_versions.c for different layout version.
    
Add tests in jump_map_place_obj.c to verify the layout.
    
Add boundry in merge_recx() to avoid the merge the recxs from
different stripes to mess up the rebuild.
    
Enable daos layout upgrade test in CI.
    
Required-githooks: true
    
Features: pr rebuild ec
Skip-build-ubuntu20-rpm: true
Allow-unstable-test: true
Signed-off-by: Di Wang <di.wang@intel.com>
* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/daos_test/suite.py [hunk 1]
```diff
@@ -408,3 +408,17 @@ def test_daos_dedup(self):
         :avocado: tags=daos_test,daos_core_test,test_daos_dedup
         """
         self.run_subtest()
+    def test_daos_upgrade(self):
```
Comment (daltonbohning) [inline src/tests/ftest/daos_test/suite.py:411]: There should be a newline before this definition

File: src/tests/ftest/daos_test/suite.py [hunk 2]
```diff
@@ -408,3 +408,17 @@ def test_daos_dedup(self):
...
+        """Jira ID: DAOS-1568
+
+        Test Description:
+            Run daos_test -P
+
+        Use cases:
+            Core tests for daos_test
+
+        :avocado: tags=all,pr,daily_regression
+        :avocado: tags=hw,ib2,medium
```
Comment (daltonbohning) [inline src/tests/ftest/daos_test/suite.py:421]: ib2 is no longer used
```suggestion
        :avocado: tags=hw,medium
```

File: src/tests/ftest/daos_test/suite.py (hunk 1 above)
Comment (wangdi1) [reply src/tests/ftest/daos_test/suite.py:411]: sure

File: src/tests/ftest/daos_test/suite.py (hunk 2 above)
Comment (wangdi1) [reply src/tests/ftest/daos_test/suite.py:421]: sure, thanks

File: src/tests/ftest/daos_test/suite.py [hunk 3]
```diff
@@ -408,3 +408,18 @@ def test_daos_dedup(self):
         :avocado: tags=daos_test,daos_core_test,test_daos_dedup
         """
         self.run_subtest()
+
+    def test_daos_upgrade(self):
+        """Jira ID: DAOS-1568
```
Comment (daltonbohning) [inline src/tests/ftest/daos_test/suite.py:413]: Maybe should be DAOS-10681?

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM except minor ticket reference

File: src/tests/ftest/daos_test/suite.py (hunk 3 above)
Comment (wangdi1) [reply src/tests/ftest/daos_test/suite.py:413]: ah, sure. though it seems all test cases here are using DAOS-1568.

Comment (daltonbohning) [reply src/tests/ftest/daos_test/suite.py]: Yeah, probably just copy-paste :)

**Review by jolivier23 — DISMISSED** [review]

**Comment by wangdi1** [comment]
all failures are known and unrelated failure. ping reviewers. Thanks

**Comment by wangdi1** [comment]
I split this patch into a few smaller patch for reviewing purpose. let's close this one.

**Pull request closed without merging**